### ISSUE TITLE
Grid manipulation code?

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [pyugrid/pyugrid#137]
A discussion on #135 got side tracked onto what code we should have for grid manipulation, and where should we get it. So I'm starting this one to capture that better.

Highlights:
PR: #111 uses cell_tree2d (https://github.com/NOAA-ORR-ERD/cell_tree2d) to do point in-cell-location. That is kept as a separate project, because it's useful for py_sgrid, and potentially other projects as well. cell_tree2d is a high performance locator written in C++, and wrapped in Python/numpy. It's new, but seems to be working well and is fast and robust and easy to build. We will in all likelihood use it operational in our GNOME project, either with or without py_ugrid, so it's likely to be well maintained.

As it turns out, there is more triangular mesh grid handling code in MPL that we were aware of. in particular, there is a trapezoidal-map implementation for point-in-cell location. With a quick, non exhaustive test, it appears to have similar performance as cell_tree2d 

The code in MPL is a bit buried in MPL -- if we want to use it (or some parts of it, it would probably make sense to pull it out into a separate project --- it is open-source after all. There was some talk of a triangular grid lib -- I think maybe that should go into, or be merged with cell_tree-2d, but we can start with some stuff being buried in py_ugrid, too.

So as we move forward, if grid-manipulation functionality is need, it's good idea to look in MPL to see if it's there already.

I propose we take a look at the MPL code to see if it has functionality we need, and go borrow it if it does (ditto for SciPy).

@hetland wrote: """Also, there is quite a bit of Triangle stuff in scipy now. I expect qhull
is a great way to do things."""

I'm pretty sure that the MPL triangle code also uses qhull -- anyway, yes, that could be useful, doesn't do constrained Delaunay, unfortunately :-(

Final note: ugrids and curvilinear grids need some of the same stuff, It wouldn't be a good idea to be too focused on only triangular meshes. In fact, that's one good reason for us not to dump cell_tree2d in favor of the stuff in MPL -- it supports  quad grids too.

Pinging a few folks to make sure this gets around:
@svater, @ocefpaf, @jay-hennen, @rhattersley, @rsignell-usgs

**Comment by jay-hennen** [comment]
When I look at the mpl triangulation API, these are my impressions:
http://matplotlib.org/api/tri_api.html

`class matplotlib.tri.Triangulation(x, y, triangles=None, mask=None)`
Basically the same as a 'dumb' pyugrid. It represents the grid and nothing else.

`class matplotlib.tri.TriFinder(triangulation)`
Same functionality as `ugrid.locate_faces()`, but as a separate class. TrapezoidMapTriFinder performance is usually within +-30% the speed of CellTree, though this difference can be much larger on edge cases.

`class matplotlib.tri.TriInterpolator(triangulation, z, trifinder=None)`
same functionality as `ugird.interpolate_var_to_points()`, but as a separate class for each type of interpolation, sort of like an Iris cube that only has `interpolate()` available. Personally, I find this to be a very annoying way of presenting this functionality. Doing it like this is intuitive and potentially cleaner, but also makes designing an API that allows short-circuiting of the interpolation much more confusing. If I want to interpolate five variables to the same points (or, interpolate 2000 slices of the same variable to the same points, I do NOT want to calculate the interpolation weights 5 or 2000 times. However, many people don't care about all that and just want to interpolate the damn variable. One example of this is API complexity is in https://github.com/sgrid/pysgrid/pull/85.

Frustratingly enough, I might end up building something that tastes just like a `iris.cube` for GNOME soon, and have to figure out a way to avoid all that complexity migrating up the stack.

`class matplotlib.tri.UniformTriRefiner(triangulation)`
Re-grids a triangulation and a variable associated with it and provides the new triangulation and variable.

**TLDR:** Mostly stuff we've already re-implemented in some way, or need a much more nuanced API for (interpolation)
